### PR TITLE
SI-7747 Make REPL wrappers serialization friendly.

### PR DESCRIPTION
**Pull request by ScrapCodes**
Re - submit of #4522. 

Includes more tests and incorporated discussion from the source PR.

### CODE REVIEW

**Comment by retronym** [comment]
LGTM

**Comment by som-snytt** [comment]
To coin a phrase, any chance of getting this committed?

Green checks, looks green to me, etc.

**Comment by som-snytt** [comment]
Tack!

File: test/files/run/repl-serialization.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      val out = new ObjectOutputStream(bos)
+      out.writeObject(o)
+      out.close()
+      bos.toByteArray
+    }
+    println("== evaluating lambda")
+    extract.value.asInstanceOf[() => Any].apply()
+    println("== reconstituting into a fresh classloader")
+    val reconstituted = deserializeInNewLoader(serialize(extract.value)).asInstanceOf[() => Any]
+    println("== evaluating reconstituted lambda")
```
Comment (prashant7090) [inline test/files/run/repl-serialization.scala:65]: Can we use logger instead of println()?

**Comment by ScrapCodes** [comment]
I just tried they merge cleanly on 2.12.x, should we make a separate PR for 2.12.x or this might go along with the periodical update of 2.12.x branch ? I am just not aware of the mechanics.

**Comment by som-snytt** [comment]
@ScrapCodes  PRs merge forward.

File: test/files/run/repl-serialization.scala (hunk 1 above)
Comment (som-snytt) [inline test/files/run/repl-serialization.scala:65]: @prashant7090 Those "run" tests produce output that is diff'd against expected output. Logger output doesn't make sense unless the check file expresses "info" output versus "error", etc, which is an interesting idea but not how the test runner works right now. It is able to ifdef against test settings, but the mechanism is fragile. The check file could declare that contains logging at a certain level, and the runner could filter.

Comment (prashant7090) [inline test/files/run/repl-serialization.scala:65]: ok Thanks!

**Pull request merged**